### PR TITLE
Generate tests: modality-agnostic input preparation

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Requirement for #33212
Follow-up to #33663  

This PR rewrites the LLM-centric `_get_input_ids_and_config()`, the function that creates random model inputs for tests, into a modality-agnostic `prepare_config_and_inputs_for_generate()`

The rest of the diff consists in propagating the change. Highlights:
1. most `_get_input_ids_and_config()` overwrites were deleted as a result of the changes 🔪 
2. most test generate calls receive a dictionary of input, as opposed to `input_ids`
3. `_check_outputs` now receives the model's main input, as opposed to `input_ids`
4. because of the changes above, a few test overwrites that needed to be updated could be deleted instead 🙏 

___________________________

In a follow-up PR: hunt tests that no longer need to be skipped/overwritten as a result of these changes 🎯

### CODE REVIEW

File: tests/generation/test_utils.py [hunk 1]
```diff
@@ -2088,15 +2002,20 @@ def test_inherits_generation_mixin(self):
...
         config = config.text_config if hasattr(config, "text_config") else config
         num_sequences_in_output = batch_size * num_return_sequences
 
         gen_len = (
             output.sequences.shape[-1] - 1 if config.is_encoder_decoder else output.sequences.shape[-1] - seq_length
         )
 
+        # in some models we subsample the sequence length in inner layers
+        if hasattr(self.model_tester, "get_subsampled_output_lengths"):
+            seq_length = self.model_tester.get_subsampled_output_lengths(seq_length)
```
Comment (gante) [inline tests/generation/test_utils.py:2017]: some models were overwritting `_check_outputs` to apply subsampling on the sequence length. Since it was a single pattern in common with the overwrites, and overloading couldn't be applied here (it changes the internals), I've decided to move the pattern here

File: tests/models/chameleon/test_modeling_chameleon.py [hunk 2]
```diff
@@ -116,7 +116,7 @@ def prepare_config_and_inputs(self):
 
         input_mask = None
         if self.use_input_mask:
-            input_mask = torch.tril(torch.ones(self.batch_size, self.seq_length)).to(torch_device)
+            input_mask = torch.tril(torch.ones_like(input_ids).to(torch_device))
```
Comment (gante) [inline tests/models/chameleon/test_modeling_chameleon.py:119]: common pattern: `input_mask`, which was then passed around as `attention_mask`, was a `torch.float32` and not a `torch.long` 👀

File: tests/models/musicgen/test_modeling_musicgen.py [hunk 3]
```diff
@@ -124,6 +120,7 @@ def __init__(
         pad_token_id=99,
         bos_token_id=99,
         num_codebooks=4,
+        audio_channels=1,
```
Comment (gante) [inline tests/models/musicgen/test_modeling_musicgen.py:123]: `audio_channels=1` is the default in the config. In other words, this doesn't change the tests, but allows us to quickly override the config (see overloaded test below)

File: tests/models/musicgen/test_modeling_musicgen.py [hunk 4]
```diff
@@ -1415,170 +1385,10 @@ def test_model_get_set_embeddings(self):
             lm_heads = model.get_output_embeddings()
             self.assertTrue(lm_heads is None or isinstance(lm_heads[0], torch.nn.Linear))
 
-    def _get_input_ids_and_config(self, batch_size=2):
```
Comment (gante) [inline tests/models/musicgen/test_modeling_musicgen.py:1418]: (all this deleted code corresponds to overwritten functions that no longer need to be overwritten)

File: tests/generation/test_utils.py [hunk 5]
```diff
@@ -97,41 +97,37 @@ class GenerationTesterMixin:
...
 
-        if self.has_attentions:
-            attention_mask = torch.ones_like(input_ids, dtype=torch.long)
-        else:
-            attention_mask = None
-
-        # It is important set set the eos_token_id to None to ensure that no sequences
-        # shorter than `max_length` can be generated
+        # It is important set `eos_token_id` to `None` to avoid early stopping (would break for length-based checks)
+        if config.eos_token_id is not None and config.pad_token_id is None:
```
Comment (zucchini-nlp) [inline tests/generation/test_utils.py:123]: i just had realized, does this work for composite models or we also need to `config.get_text_config().eos_token_id`? Just a thought, since I started noticing some patterns where modeling code expects the text config to be modified

At least it was the case for `config.output_attentions`

File: tests/generation/test_utils.py [hunk 6]
```diff
@@ -440,28 +401,26 @@ def _contrastive_generate(
     @pytest.mark.generate
     def test_greedy_generate(self):
         for model_class in self.all_generative_model_classes:
-            config, input_ids, attention_mask, inputs_dict = self._get_input_ids_and_config()
+            config, inputs_dict = self.prepare_config_and_inputs_for_generate()
+            main_input = inputs_dict[self.input_name]
```
Comment (zucchini-nlp) [inline tests/generation/test_utils.py:405]: wondering what happens if we use `model_cls.main_input_name`? AFAIR from few months ago, there were some inconsistencies in how model main input is defined, and we could make another round of cleaning up on that because `main_input_name` is also used in `generate()`. Maybe we can have a more generalized interface and testing suite?

File: tests/generation/test_utils.py [hunk 7]
```diff
@@ -1328,9 +1253,8 @@ def test_dola_decoding_sample(self):
                 "use_cache": hasattr(config, "use_cache"),  # Some models don't support the cache
             }
             generation_kwargs.update({"dola_layers": "low"})
-            model_kwargs = {"attention_mask": attention_mask} if attention_mask is not None else {}
-            output_dola = model.generate(input_ids, **model_kwargs, **generation_kwargs, **inputs_dict)
-            self._check_outputs(output_dola, input_ids, model.config, use_cache=hasattr(config, "use_cache"))
+            output_dola = model.generate(**generation_kwargs, **inputs_dict)
+            self._check_outputs(output_dola, main_input, model.config, use_cache=hasattr(config, "use_cache"))
```
Comment (zucchini-nlp) [inline tests/generation/test_utils.py:1257]: totally unrelated to this PR: why is it `hasattr(config, "use_cache")` and not `getattr`?

File: tests/models/musicgen/test_modeling_musicgen.py (hunk 4 above)
Comment (zucchini-nlp) [reply tests/models/musicgen/test_modeling_musicgen.py:1418]: love it!

File: tests/models/reformer/test_modeling_reformer.py [hunk 8]
```diff
@@ -684,20 +684,15 @@ def _check_hidden_states_for_generate(
...
-        _ = inputs_dict.pop("attention_mask", None)
-        _ = inputs_dict.pop("decoder_input_ids", None)
-        _ = inputs_dict.pop("decoder_attention_mask", None)
-        input_ids = input_ids[:batch_size, :16]
-        attention_mask = torch.ones_like(input_ids, dtype=torch.long)[:batch_size, :16]
-        config.eos_token_id = None
-        config.forced_eos_token_id = None
-        return config, input_ids, attention_mask, inputs_dict
+        original_sequence_length = self.model_tester.seq_length
+        self.model_tester.seq_length = 16
```
Comment (zucchini-nlp) [inline tests/models/reformer/test_modeling_reformer.py:692]: nice to see a cleaner way

**Review by zucchini-nlp — APPROVED** [review]
Awesome! So much code cleaned up, thanks! 💓 

Overall looks good to me, just a few question for my general understanding. I see some VLMs are failing the CI. I remember skipping one of the beam search tests for VLMs earlier so it's prob that. But lmk if you want me to look at it :)

File: tests/models/musicgen/test_modeling_musicgen.py (hunk 4 above)
Comment (LysandreJik) [reply tests/models/musicgen/test_modeling_musicgen.py:1418]: That's very nice

**Review by LysandreJik — APPROVED** [review]
Very welcome and clean changes! Nice to see less tests being overwritten :ok_hand:

**Review by ArthurZucker — APPROVED** [review]
Very welcome.

IMO we should have:
- a dict of common forward input names, with default values, that range from 0 to the small model vocab size. 
- a variant for vLMs
- a variant for image only 
- a variant for audio 
And then you just take them using inspect to inspect the forward pass. 

Now for testing the padding and else, of course you need somethibng else, but this way people don't have to do this ever again: https://github.com/huggingface/transformers/blob/1dba608df93ffb10a9c268ef35191adf2424c5ca/tests/models/sam/test_modeling_sam.py#L448-L458

File: tests/generation/test_utils.py (hunk 5 above)
Comment (gante) [reply tests/generation/test_utils.py:123]: Good point! I agree it should be set in the text config, will update

File: tests/generation/test_utils.py (hunk 7 above)
Comment (gante) [reply tests/generation/test_utils.py:1258]: 👀 good catch, will change

**Comment by gante** [comment]
@ArthurZucker that is a cool idea, model-agnostic inputs, potentially dependent on modality (or perhaps simply looking at the signature of the fwd pass?)

I took note of it to explore after the current round of reactors, to avoid adding more parallel threads :D

File: tests/generation/test_utils.py (hunk 6 above)
Comment (gante) [reply tests/generation/test_utils.py]: Not only it worked, but also it allowed us to remove the `input_name` attribute from all testers 💛
